src/range.cr

```diff
@@ -305,4 +305,46 @@ struct Range(B, E)
       self
     end
   end
+
+  #Perform binary search
+  #Based on Marc-André Lafortune's Ruby backports implementation, ported by T. Yamada
+  def bsearch
+    from = self.begin
+    to   = self.end
+    unless from.is_a?(Int) && to.is_a?(Int)
+      # Float support is currently dropped
+      raise "can't do binary search for #{from.class}"
+    end
+
+    midpoint = 0 # placeholder
+    convert = ->{ midpoint }
+
+    to -= 1 if excludes_end?
+    satisfied = nil
+    while from <= to
+      midpoint = (from + to)/2
+      result = yield(cur = convert.call)
+      case result
+      when Int
+        return cur if result == 0
+        result = result < 0
+      when Float
+        return cur if result == 0
+        result = result < 0
```

Comment:
The compiler doesn't support `is_a?(...) || is_a?(...)` yet (which is what case/when is rewritten to) so you'll have to put separate cases with same code (or a helper method) for now.